### PR TITLE
Fix issue 8370

### LINKED ISSUE #8370: NumbaTypeError: Cannot modify readonly array of type: readonly array(int64, 2d, C)
```
import numpy as np
from numba import njit

@njit
def unfold(x, y):
    r, c = x.shape
    a = np.broadcast_to(x, (y, r, c))
#     a = a.copy()
    b = np.swapaxes(a, 0, 1)
#     b = b.copy()
    cc = b.flatten()
#     c = c.copy()
    d = np.reshape(cc, (-1, c))
    # assign indicator to the right end of the period
    # in case of look forward bias
#     d = d.copy()
    d[y - 1 :, :] = d[: 1 - y]
    return d


unfold(np.array([[1, 1, 1], [2, 2, 2], [3,3,3]]), 2)
```
This will reproduce the error above. Under plain numpy this works fine.
By uncommenting any one of the copy() calls, the issue is resolved. So I have no idea which numpy call returns this readonly array.
Can it be implemented without an explicit copy call?

### PR DESCRIPTION
**Pull request by bszollosinagy**
Fixes #8370 

New unit tests were created, and the solution added, as suggested by @gmarkall 

All unit tests pass locally that would exercise ravel() and flatten()

- python -m numba.runtests -m -v numba.tests.test_array_reductions.TestArrayReductions
- python -m numba.runtests -m -v numba.tests.test_array_manipulation.TestArrayManipulation
- python -m numba.runtests -m -v numba.tests.test_closure.TestObjmodeFallback
- python -m numba.runtests -m -v numba.tests.test_dyn_array.TestNpStack
- python -m numba.runtests -m -v numba.tests.test_np_functions.TestNPFunctions.test_extract_basic
- python -m numba.runtests -m -v numba.tests.test_support.TestAssertPreciseEqual

CUDA tests may need to be run, but I simply cannot download the CUDA drivers from Nvidia today for some reason, so maybe a CI test would be more appropriate.

### CODE REVIEW

**Comment by gmarkall** [comment]
Many thanks for the PR!

> CUDA tests may need to be run, but I simply cannot download the CUDA drivers from Nvidia today for some reason, so maybe a CI test would be more appropriate.

I'm not sure `flatten` would work on CUDA since its implementation returns a copy in Numba, and it's not possible to do allocation for the CUDA target - am I missing something here?

**Comment by gmarkall** [comment]
gpuci run tests

**Comment by gmarkall** [comment]
(Marking as RFR because the test fail is a CI issue, not an issue with this PR)

File: numba/core/typing/arraydecl.py
Comment (stuartarchibald) [inline numba/core/typing/arraydecl.py:477]: I'm not sure that this is correct? I think the mutability of the array resulting from `ravel()` would be determined by whether a copy is needed?
```pycon
In [35]: import numpy as np

In [36]: a = np.ones((2, 3))

In [37]: a.flags.writeable = False

In [38]: a.flags
Out[38]: 
  C_CONTIGUOUS : True
  F_CONTIGUOUS : False
  OWNDATA : True
  WRITEABLE : False
  ALIGNED : True
  WRITEBACKIFCOPY : False
  UPDATEIFCOPY : False

In [39]: a_c = np.ravel(a)

In [40]: a_c.flags
Out[40]: 
  C_CONTIGUOUS : True
  F_CONTIGUOUS : True
  OWNDATA : False
  WRITEABLE : False
  ALIGNED : True
  WRITEBACKIFCOPY : False
  UPDATEIFCOPY : False

In [41]: a_f = np.ravel(a, "F")

In [42]: a_f.flags
Out[42]: 
  C_CONTIGUOUS : True
  F_CONTIGUOUS : True
  OWNDATA : True
  WRITEABLE : True
  ALIGNED : True
  WRITEBACKIFCOPY : False
  UPDATEIFCOPY : False

In [43]: a.ravel('F').flags
Out[43]: 
  C_CONTIGUOUS : True
  F_CONTIGUOUS : True
  OWNDATA : True
  WRITEABLE : True
  ALIGNED : True
  WRITEBACKIFCOPY : False
  UPDATEIFCOPY : False

In [44]: a.ravel('C').flags
Out[44]: 
  C_CONTIGUOUS : True
  F_CONTIGUOUS : True
  OWNDATA : False
  WRITEABLE : False
  ALIGNED : True
  WRITEBACKIFCOPY : False
  UPDATEIFCOPY : False
```

Comment (gmarkall) [reply numba/core/typing/arraydecl.py:477]: Note that in Numba, it appears that no copy is made if the array is C-contiguous, and a copy is made if it is not C-contiguous: https://github.com/numba/numba/blob/main/numba/np/arrayobj.py#L1992

Comment (bszollosinagy) [reply numba/core/typing/arraydecl.py]: Okay, updated the code to be sensitive to this, and added a unit test to verify.

**Comment by bszollosinagy** [comment]
Resolved issue picked up by an automated test: if the source array is mutable, then a simple ravel() should not make it read-only.

**Comment by bszollosinagy** [comment]
Ready for review.

**Comment by bszollosinagy** [comment]
Note: the algo modification is not a big deal: 

- on calling flatten(), a copy is made by Numba, and therefore the read-only is removed (just like in Numpy, see ref 1).  
- In the case of ravel(), the copy is only made if the "order" of the dimensions has to be converted from Fortran to C, otherwise no copy is made. 
  - In case there was a copy, then the read-only is removed (just like in Numpy, see ref 1),
  - in case no copy was needed, then we look at the source array: was it read-only.

That's all.

ref: [Numpy removing writeable=False when doing a copy()](https://numpy.org/devdocs/reference/generated/numpy.copy.html)

**Comment by gmarkall** [comment]
gpuci run tests

File: numba/tests/test_array_manipulation.py
Comment (gmarkall) [inline numba/tests/test_array_manipulation.py:1268]: This isn't the message I get - I get:

```
No implementation of function Function(<built-in function setitem>) found for signature:
 
 >>> setitem(readonly array(int64, 2d, C), UniTuple(slice<a:b> x 2), readonly array(int64, 2d, C))
```

I think it'd be OK to just delete this comment rather than update it.

File: numba/tests/test_array_manipulation.py
Comment (gmarkall) [inline numba/tests/test_array_manipulation.py:1287]: Similar to above, I think this comment can / should be removed.

File: numba/core/typing/arraydecl.py
Comment (gmarkall) [inline numba/core/typing/arraydecl.py:486]:
```suggestion
        # Just like in NumPy: "np.copy clears previously
```

File: numba/core/typing/arraydecl.py
Comment (gmarkall) [inline numba/core/typing/arraydecl.py:488]: I'm unclear about what this comment is saying. Is there any deviation from NumPy here, or do Numba and NumPy have similar behaviour now?

**Review by gmarkall — CHANGES_REQUESTED** [review]
Many thanks for the updates and your continued efforts! There are a couple of comments on the PR.

In addition, in `test_readonly_after_ravel` and `test_readonly_after_flatten`, there is no assertion that checks that an expected output is produced - could you also assert that `cfunc` and `pyfunc` both produce the same return value please?

**Comment by gmarkall** [comment]
Also, apologies for the delay in this round of review.

File: numba/core/typing/arraydecl.py
Comment (bszollosinagy) [reply numba/core/typing/arraydecl.py:488]: Numba and NumPy will have identical behaviour. The comment explains why the next line has a hardcoded readonly=False  argument.

Comment (bszollosinagy) [reply numba/core/typing/arraydecl.py:491]: I've now rephrased the comment.

File: numba/tests/test_array_manipulation.py
Comment (gmarkall) [inline numba/tests/test_array_manipulation.py:1336]: If you use `np.testing.assert_array_equal(res_py, res_nb)` here instead of `self.assertTrue(np.array_equal(...`, then if the test fails, NumPy outputs the difference.

E.g. for a trivial example:

```
In [2]: np.testing.assert_array_equal(1, 2)

# ... snip ...

AssertionError: 
Arrays are not equal

Mismatched elements: 1 / 1 (100%)
Max absolute difference: 1
Max relative difference: 0.5
 x: array(1)
 y: array(2)
```

**Review by gmarkall — CHANGES_REQUESTED** [review]
Many thanks for the quick update - just a very small comment on the diff - would you be happy to make this change, then this should be good to approve?

File: numba/tests/test_array_manipulation.py
Comment (bszollosinagy) [reply numba/tests/test_array_manipulation.py]: sounds good, done!

**Comment by gmarkall** [comment]
Many thanks - just waiting on CI before I approve.

**Comment by gmarkall** [comment]
gpuci run tests

**Review by gmarkall — APPROVED** [review]
Many thanks for the PR and updates! I think this looks good now.

**Pull request merged**
